Update terraform tutorial docs (#39377)
gravitational/teleport: docs/pages/management/dynamic-resources/terraform-provider.mdx

Reviewer: It feels like this guide is now pretty heavily overlapped with `/machine-id/access-guides/terraform.mdx` now that we're only recommending Machine ID be used to create credentials for the Terraform provider - I think it'd be useful to distinguish or merge these. WDYT @ptgott

This guide demonstrates how to: Set up the Terraform provider for Teleport on
Linux and macOS.

> Terraform is intended to be used with Machine ID, Teleport's feature for identifying

Reviewer: ```suggestion
Terraform is intended to be used with Machine ID, Teleport's feature for providing identities to
```

## Step 1/3. Create Teleport credentials for Terraform

Terraform needs a signed identity file from the Teleport cluster certificate
authority to manage resources in the cluster. Terraform is intended to be used with
[(Machine ID)](../../machine-id/introduction.mdx), so this tutorial will show you how to

> use a local Machine ID bot to try it out.

Reviewer 2: I think this information about the scope of the guide (i.e., that it includes Machine ID instructions) belongs in the introduction of the guide.

Author: Updated!

For production, Terraform should be run using Machine ID on a CI/CD platform.
For more information on using Machine ID and Terraform in a production setting, see the

> [Machine ID Terraform Guide](../../machine-id/access-guides/terraform.mdx).

Reviewer 2: I would move this to "Next steps" so the reader can sprint through the tutorial first before planning work for production.

Author: Updated!

1. Create a new file called `terraform.yaml` and open it in an editor.

> 1. Configure the necessary role by pasting the following content into a `terraform_role.yaml` file:

Reviewer 2: I would clarify what entity will have this role and what it uses the role for

Author: Done, let me know what you think of the new language.

           - access_list
           verbs: ['list','create','read','update','delete']
   version: v7
   ```

   These settings configure a role named `terraform` with the permissions
   required to manage resources in your Teleport cluster.

1. Create the `terraform` role by running the following command:

   ```code
   $ tctl create -f terraform_role.yaml
   role 'terraform' has been created
   ```

1. We're going to create two scripts that will let us quickly and repeatably create and destroy
   our local Terraform bot. First, create a new file called `create_and_run_terraform_bot.sh`,
   open it in an editor, and paste in the following content:

   ```shell
   #! /usr/bin/env bash

   TOKEN=$(tctl bots add terraform-test --roles=terraform --format=json | jq -r '.token_id')
   tbot start \
     --destination-dir=./terraform-identity \
     --token="$TOKEN" \
     # Replace with the host and port of your Teleport Auth Service
     # or Teleport Proxy Service
     --auth-server=tele.example.com:443 \
     --join-method=token
   ```

   When run, this creates a new bot with the token retrieved from the JSON output. That token

>    is piped into the start command. For security reasons, we don't allow bot join tokens to be

Reviewer: > For security reasons, we don't allow bot join tokens to be created in any way other than adding a bot

I'm not sure this is true - if anything it's an awkward edge in the UX experience at the moment that should be fixed in coming months.

Author: Ok, I remember seeing/hearing something that it was for security. Maybe that was Ben making it up too hah.

           - access_list
           verbs: ['list','create','read','update','delete']
   version: v7
   ```

   These settings configure a role named `terraform` with the permissions
   required to manage resources in your Teleport cluster.

1. Create the `terraform` role by running the following command:

   ```code
   $ tctl create -f terraform_role.yaml
   role 'terraform' has been created
   ```

> 1. Create a new file called `create_and_run_terraform_bot.sh`, open it in an editor, and

Reviewer 2: I think it would make the guide a little simpler to turn the shell scripts into commands that the reader would paste directly into the terminal at the appropriate pointa in the guide, similar to the way we do it in other docs guides. That way, they wouldn't need to manage files and file permissions.

**Edit:** After finishing the guide, I saw the note that you can re-run the two bot scripts to start/stop the bot. If we're keeping the shell scripts, maybe we could add this clarification earlier in the guide?

Author: Added context here. Yeah, this is just a [not friendly process](https://gravitational.slack.com/archives/C08HF8E9F/p1707942263917899), unfortunately.

paste in the following content:

   ```shell
   #! /usr/bin/env bash

   set -euxo pipefail

>    TOKEN=$(tctl bots add terraform-test --roles=terraform --format=json | jq -r '.token_id')

Reviewer 3: Side note: `tctl tokens add` has a `--format=token` option that just outputs the raw token, making `jq` unnecessary here. We should probably do the same for `tctl bots add` at some point.

Author: @zmb3 I got this when trying to use `--format=token`

```
ERROR: enum value must be one of text,json, got 'token'
```

Author: Oh you're saying `tokens add`, this is `bots add`. I've been through the rigamorole that ended in "you can't create a token and then give it to the bot as a value.

   tbot start \
     --destination-dir=./terraform-identity \
     --token="$TOKEN" \

>      --auth-server=teleport-ent-15.asteroid.earth:443 \

Reviewer 2: I would clarify that the user should edit this flag

Reviewer 3: I would also use tele.example.com:443 since that's what we've used in other places in this PR.

Author: Good catch, that's copy pasta. Thank you.

    --join-method=token
  ```

  When run, this creates a new bot with the token retrieved from the JSON output. That token
  is piped into the start command. For security reasons, we don't allow bot join tokens to be
  created in any way other than adding a bot, so for local development work, you must manually
  manage the bot and its token.

1. Create another file called `remove_terraform_bot.sh`, open it in an editor, and
paste in the following content:

  ```shell
  #! /usr/bin/env bash

  set -euxo pipefail

  tctl bots rm terraform-dave
  rm -rf ./terraform-identity
  ```

  This script will be used later to clean up our bot in between runs and when we're done.

> 1. In a new terminal windwo, run the `create_and_run_terraform_bot.sh` script to create the new bot

Reviewer 3: ```suggestion
1. In a new terminal window, run the `create_and_run_terraform_bot.sh` script to create the new bot
```

   $ terraform apply
   ```

> 1. When you're done, use CNTL-C to stop the Terraform bot. Then run the `remove_terraform_bot.sh` script to remove the bot and clean up the `terraform-identity` folder:

Reviewer 3: ```suggestion
1. When you're done, use Ctrl-C to stop the Terraform bot. Then run the `remove_terraform_bot.sh` script to remove the bot and clean up the `terraform-identity` folder:
```

Just to match what already appears in help messages in the product.
